File: Editorial/151_Reverse_Words_in_a_String/Go/2_Reverse_the_Whole_String_and_Then_Reverse_Each_Word.go
// Translated Go solution
func reverseWords(s string) string {
    // convert string to char slice
    // and trim spaces at the same time
    arr := trimSpaces(s)

    // reverse the whole string
    reverse(arr, 0, len(arr)-1)

    // reverse each word
    reverseEachWord(arr)

    return strings.Join(arr, "")
}

func trimSpaces(s string) []string {
    left, right := 0, len(s)-1
    // remove leading spaces
    for left <= right && s[left] == ' ' {
        left++
    }

    // remove trailing spaces
    for left <= right && s[right] == ' ' {
        right--
    }

    // reduce multiple spaces to single one
    output := make([]string, 0)
    for left <= right {
        if s[left] != ' ' {
            output = append(output, string(s[left]))
        } else if output[len(output)-1] != " " {
            output = append(output, string(s[left]))
        }
        left++
    }
    return output
}

func reverseEachWord(c []string) {
    n := len(c)
    start, end := 0, 0

    for start < n {
        // go to the end of the word
        for end < n && c[end] != " " {
            end++
        }
        // reverse the word
        reverse(c, start, end-1)
        // move to the next word
        start = end + 1
        end++
    }
}

func reverse(c []string, start int, end int) {
    for start < end {
        c[start], c[end] = c[end], c[start]
        start++
        end--
    }
}

File: Editorial/151_Reverse_Words_in_a_String/C/2_Reverse_the_Whole_String_and_Then_Reverse_Each_Word.c
// Translated C solution
struct Node {
    char word[1000];
    struct Node* next;
};

void reverse(char* begin, char* end) {
    char temp;
    while (begin < end) {
        temp = *begin;
        *begin++ = *end;
        *end-- = temp;
    }
}

void add(struct Node** head_ref, char* new_word) {
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    strcpy(new_node->word, new_word);
    new_node->next = (*head_ref);
    (*head_ref) = new_node;
}

char* reverseWords(char* s) {
    int i = 0, j = 0;
    char temp[1000];
    struct Node* words = NULL;

    while (1) {
        if (s[i] == ' ' || s[i] == '\0') {
            if (j != 0) {
                temp[j] = '\0';
                add(&words, temp);
                j = 0;
            }

            if (s[i] == '\0') break;
        } else {
            temp[j] = s[i];
            j++;
        }
        i++;
    }

    char* ans = (char*)malloc(10000 * sizeof(char));
    struct Node* temp_node = words;
    int k = 0;

    while (temp_node != NULL) {
        for (int m = 0; m < strlen(temp_node->word); m++)
            ans[k++] = temp_node->word[m];
        if (temp_node->next != NULL)
            ans[k++] = ' ';
        else
            ans[k++] = '\0';

        temp_node = temp_node->next;
    }

    return ans;
}
